### DISCUSSION TITLE
Issue with Chem.MolFromXYZFile() while reading structures involving N(=O)[O] and/or [N+](=O)[O-] similars !

### DISCUSSION THREAD

**Discussion by amolpatil1983** [rdkit/rdkit#7918] [General] [answered]
Hi, I am not sure if this has been discussed here earlier. If anyone knows kindly direct me to the right resources.

I could successfully convert smiles "N(=O)[O] " to an xyz file using Chem.MolToXYZFile(), however while reading back the same rdkit-generated .xyz file fails although the generated .xyz file is alright.

Seemingly the conversion from any .xyz files containing N(=O)[O] or similar molecules fails. 

For example:

from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit.Chem import rdDetermineBonds

sml=Chem.CanonSmiles("N(=O)[O]")
mol=Chem.MolFromSmiles(sml)
mol=Chem.AddHs(mol)
AllChem.EmbedMolecule(mol)
Chem.MolToXYZFile(mol,"no2.xyz")

mol = Chem.MolFromXYZFile("no2.xyz")
rdDetermineBonds.DetermineBonds(mol) # Toggling allowChargedFragments =  False produces "[O][N][O]" but then misses charge info
sml=Chem.MolToSmiles(mol)
sml=Chem.CanonSmiles(sml)

Outputs:

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[53], line 12
      9 Chem.MolToXYZFile(mol,"no2.xyz")
     11 mol = Chem.MolFromXYZFile("no2.xyz")
---> 12 rdDetermineBonds.DetermineBonds(mol) # Toggling allowChargedFragments =  False produces "[O][N][O]" but then misses charge info
     13 sml=Chem.MolToSmiles(mol)
     14 sml=Chem.CanonSmiles(sml)

ValueError: Final molecular charge does not match input; could not find valid bond ordering

Any ideas?

**Comment by greglandrum** [comment, marked as answer]
@amolpatil1983 the problem here is that the molecule you are providing includes a radical on the oxygen atom. This is not captured in the output XYZ file, so you have a non-stable structure in the XYZ file. `DetermineBonds()` fails in these cases.

If you describe a nitro group correctly everything works. Here's an example for nitro-methane:
```
In [3]: m = Chem.MolFromSmiles('C[N+](=O)[O-]')

In [4]: m = Chem.AddHs(m)

In [5]: rdDistGeom.EmbedMolecule(m,randomSeed=42)
Out[5]: 0

In [6]: xyz = Chem.MolToXYZBlock(m)

In [7]: nm = Chem.MolFromXYZBlock(xyz)

In [8]: from rdkit.Chem import rdDetermineBonds

In [11]: rdDetermineBonds.DetermineBonds(nm)

In [12]: nm.Debug()
Atoms:
        0 6 C chg: 0  deg: 4 exp: 4 imp: 0 hyb: SP3
        1 7 N chg: 1  deg: 3 exp: 4 imp: 0 hyb: SP2
        2 8 O chg: 0  deg: 1 exp: 2 imp: 0 hyb: SP2
        3 8 O chg: -1  deg: 1 exp: 1 imp: 0 hyb: SP2
        4 1 H chg: 0  deg: 1 exp: 1 imp: 0 hyb: S
        5 1 H chg: 0  deg: 1 exp: 1 imp: 0 hyb: S
        6 1 H chg: 0  deg: 1 exp: 1 imp: 0 hyb: S
Bonds:
        0 1->0 order: 1
        1 2->1 order: 2 conj?: 1
        2 3->1 order: 1 conj?: 1
        3 4->0 order: 1
        4 5->0 order: 1
        5 6->0 order: 1
```

**Reply by amolpatil1983** [reply]
Thanks for the clarification, sir !
BTW: Thanks are also due to you for bringing out such a nice program package !
